community/kernel/src/main/java/org/neo4j/kernel/lifecycle/LifeSupport.java

```diff
@@ -307,6 +307,8 @@ public synchronized void restart( Lifecycle instance )
     {
         if ( instance instanceof Lifecycle )
         {
+            checkContains( (Lifecycle) instance );
```

Comment:
I love this, this should be a great help in finding issues around lifecycle management.

Since how we wire things up is mostly static though, perhaps we should only do this when assertions are enabled?